### ISSUE TITLE
Increase default f0max for TTS, and Praat is a better pitch extractor than Dio?

### ISSUE THREAD

**Issue by iamanigeeit** [espnet/espnet#5259]
Hello,

400 Hz is fine for read speech, but not suitable for expressive / emotional TTS applications. For example, 
[0001_001491](https://github.com/espnet/espnet/files/11870982/0001_001491.txt) (please rename to .wav as Github doesn't allow .wav uploads) from the [Emotional Speech Dataset](https://hltsingapore.github.io/ESD/) has 500+ Hz segments according to Praat, and some other samples even have 800+ Hz.

Praat (you can play back the extracted pitch contour and verify the pink dots are accurate)
![image](https://github.com/espnet/espnet/assets/13131208/4eb12bd0-470a-41b3-a9f5-59735d9701c5)

Pitch (blue) extracted with Dio f0max at 400 Hz -- notice the cutoff at 400 Hz, and the missing syllables:
![image](https://github.com/espnet/espnet/assets/13131208/aea8742a-9919-43e5-ad9c-b5a2a3eee22a)

The Dio default is actually 800. Changing f0max to 800 helps, but there's some discontinuities:
![image](https://github.com/espnet/espnet/assets/13131208/d3bfceac-1cd8-476b-bdc6-8dd90d98b5b2)

I have no signal processing background, so if i'm doing something wrong, please let me know!

**Comment by iamanigeeit** [comment]
Update:
RAPT algorithm is supposedly better than Praat for f0 extraction, but the Python implementation at [pysptk](https://github.com/r9y9/pysptk/tree/master) does not work.

Running
```
from pysptk import sptk
f0s = pysptk.sptk.swipe(audio, fs=16000, hopsize=256, max=800)
```
gives a similar plot to Praat with more random points:
![image](https://github.com/espnet/espnet/assets/13131208/36a300a8-1622-412a-9b85-81f487f0351b)

Running
```
import librosa
f0s = librosa.yin(audio, fmin=50, fmax=800, sr=16000, frame_length=1024, hop_length=256)
```
also gives a similar messy plot:
![image](https://github.com/espnet/espnet/assets/13131208/b11c73f5-f684-4396-abb4-5c7a039747f1)

Fortunately Praat has a nice Python wrapper at [Parselmouth](https://github.com/YannickJadoul/Parselmouth).
```
import parselmouth
sound = parselmouth.Sound(audio, 16000)
f0s = sound.to_pitch(pitch_floor=75.0, pitch_ceiling=800.0)
f0s = [x[0] for x in f0s.selected_array]
```
![image](https://github.com/espnet/espnet/assets/13131208/cf4d251f-b8a4-4221-a606-280956e10e32)

The problem is that the number of pitch values returned is variable and depends on `pitch_floor`. This means it won't be compatible with other `AbsFeatExtract`.

**Comment by ftshijt** [comment]
Interesting findings~

We also have similar issues in singing voice synthesis as there are a wide pitch range to that cases. However, it is common that the pitch range would be different for almost every corpus, in that case, I think it might be better to add your customized range based on the recipe/dataset.

For pitch estimation algorithm, I think it might be related to specific scenarios. If you have other good proposals of pitch extractor, maybe you can create a PR for that?

**Comment by iamanigeeit** [comment]
> We also have similar issues in singing voice synthesis as there are a wide pitch range to that cases. However, it is common that the pitch range would be different for almost every corpus, in that case, I think it might be better to add your customized range based on the recipe/dataset.

Yes, i have done that. I ask to increase the default because models pretrained on range-limited datasets **cannot** be fine-tuned on expressive datasets. So it's extra work to train from scratch.

> For pitch estimation algorithm, I think it might be related to specific scenarios. If you have other good proposals of pitch extractor, maybe you can create a PR for that?

Yes, i'll evaluate Dio against Praat once i have the time. There are cases in LibriTTS where Dio fails to extract any pitch (`f0==0`) for every frame...

**Comment by iamanigeeit** [comment]
I have run Dio on every audio file in LibriTTS-100, and there are 182 files where Dio cannot extract any pitch:
[dio_failures.txt](https://github.com/espnet/espnet/files/11926377/dio_failures.txt)

These are normal audio files. When i run Praat on them, there is no problem, except for 1 very badly segmented audio  (8975_270782_000016_000003.wav).

Furthermore, Praat is 5-10x faster than Dio:
```
import parselmouth
import pyworld
import numpy as np
import soundfile as sf
hop_length = 256
def check_f0_praat(x, fs, feats_length):
    sound = parselmouth.Sound(x, fs)
    input_length = len(x)
    time_step = input_length * hop_length / (input_length + 4 * hop_length) / fs
    pitch = sound.to_pitch(time_step=time_step)
    f0 = np.array([p[0] for p in pitch.selected_array])
def check_f0_dio(audio, fs):
    f0, timeaxis = pyworld.dio(
        audio,
        fs=fs,
        f0_floor=75,
        f0_ceil=600,
        frame_period=hop_length / fs * 1000,
    )
    f0s = pyworld.stonemask(audio, f0, timeaxis, fs)
```
Short audio:
```
with open('downloads/LibriTTS/dev-clean/6319/64726/6319_64726_000012_000000.wav', 'rb') as f:
    audio, fs = sf.read(f)
feats_len = round(len(audio) / hop_length)

%timeit check_f0_praat(audio, fs, feats_len)
2.58 ms ± 30.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit check_f0_dio(audio, fs)
12.7 ms ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Long audio:
```
with open('downloads/LibriTTS/train-clean-100/19/198/19_198_000007_000000.wav', 'rb') as f:
    audio, fs = sf.read(f)
feats_len = round(len(audio) / hop_length)

%timeit check_f0_praat(audio, fs, feats_len)
7.24 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit check_f0_dio(audio, fs)
68.8 ms ± 686 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by ftshijt** [comment]
This sounds reasonable to me, but I'm not pretty sure how it would affect the pre-trained models we have. @kan-bayashi what's your thoughts on this?

**Comment by kan-bayashi** [comment]
Dio works well on high SNR samples but it failed when the sample containing noise.
So I think adding new f0 extractor that is robust to the noise is good idea.
And about pretrained model, maybe there is no problem since during the inference we do not use f0 extractor module.

**Comment by iamanigeeit** [comment]
It is strange, because these are LibriTTS-clean files.

Do you think changing pitch extractor params could be a problem for fine-tuning pretrained models?

**Comment by kan-bayashi** [comment]
> It is strange, because these are LibriTTS-clean files.

Here `clean` means really clean, e.g., recording in the studio level.
LibriTTS is relatively clean, but some samples are enough.

> Do you think changing pitch extractor params could be a problem for fine-tuning pretrained models?

It might effect a little bit, but if we get the more correct f0, that will be better effect.

**Comment by iamanigeeit** [comment]
Sorry, just checking: i notice we put `f0min` and `f0max` in `tts.sh`.

But it looks like something to put in `train.yaml` under `pitch_extract_conf:` because it's not a shared variable (`fs` or `n_shift`).

```
pitch_extract_conf:
    # f0min: 50
    # f0max: 1000
    reduction_factor: 1
    use_token_averaged_f0: true
```
